### PR TITLE
The whole damn shader system.

### PR DESCRIPTION
**Pull request by Jibzi**

### CODE REVIEW

**Comment by Jibzi** [comment]
[Implement shaders](https://trello.com/c/ItRdz0B3/3-implement-shaders)

File: scripts/ds_EffectData_NEW/ds_EffectData_NEW.gml [hunk 1]
```diff
@@ -2,16 +2,19 @@
 /// @param {enum} enum_effect The effect type.
```
Comment (nick-wilson95) [inline scripts/ds_EffectData_NEW/ds_EffectData_NEW.gml]: Add shader to the arguments in line 1. If we make it optional, we won't need to keep passing noone in.

File: scripts/ds_StunEffectData_New/ds_StunEffectData_New.gml [hunk 2]
```diff
@@ -6,4 +6,4 @@ var _num_steps = argument[0];
 var _id_caster = argument[1];
 
 
-return ds_EffectData_NEW(Enum_Effects.STUN, _num_steps, _id_caster);
+return ds_EffectData_NEW(Enum_Effects.STUN, _num_steps, _id_caster, noone);
```
Comment (nick-wilson95) [inline scripts/ds_StunEffectData_New/ds_StunEffectData_New.gml]: Once it's optional, we can remove the noone at the end

File: scripts/ds_DashEffectData_New/ds_DashEffectData_New.gml [hunk 3]
```diff
@@ -13,7 +13,7 @@ var _num_slowMultiplier = argument[3];
 var _num_slowSteps = argument[4];
 
 
-var _ds_DashEffectData_new = ds_EffectData_NEW(Enum_Effects.DASH, _num_steps, _id_caster);
+var _ds_DashEffectData_new = ds_EffectData_NEW(Enum_Effects.DASH, _num_steps, _id_caster, noone);
```
Comment (nick-wilson95) [inline scripts/ds_DashEffectData_New/ds_DashEffectData_New.gml]: Can remove last parameter

File: scripts/ds_EffectData_NEW/ds_EffectData_NEW.gml [hunk 4]
```diff
@@ -2,16 +2,19 @@
...
 /// @param {number} num_steps The number of steps.
 /// @param {id} id_caster The instance id of the caster.
+/// @param {shd} shd_shader The shader this effect should apply.
 /// @returns {ds_EffectData} The new effect data.
 var _enum_effect = argument[0];
 var _num_steps = argument[1];
 var _id_caster = argument[2];
+var _shd_shader = argument[3];
```
Comment (nick-wilson95) [inline scripts/ds_EffectData_NEW/ds_EffectData_NEW.gml]: var _shd_shader = argument_count >3 ? argument[3] : noone; etc...

File: objects/oDepthManager/Draw_0.gml [hunk 5]
```diff
@@ -3,7 +3,12 @@ for (var i = 0; i < ds_DepthGrid_GetHeight(ds_DepthGrid_depthGrid); i++) {
 	with (_id_depthObject) {
 		// We only want to draw the instances that actually have sprites.
 		if (sprite_index > -1) {
-			draw_self();
+			//Only apply combat shaders to oAbilityCasters
+			if (object_is_ancestor(object_index, oAbilityCaster)){
+				script_execute(Shader_PickShader());
```
Comment (nick-wilson95) [inline objects/oDepthManager/Draw_0.gml]: I think PickShader should be a private script on oAbilityCaster, also, we can have it return a boolean which represents whether we set the shader or not, and handle the draw and shader reset here - `AbilityCaster_TRY_SET_EFFECT_SHADER`

File: objects/oDepthManager/Draw_0.gml [hunk 6]
```diff
@@ -3,7 +3,12 @@ for (var i = 0; i < ds_DepthGrid_GetHeight(ds_DepthGrid_depthGrid); i++) {
 	with (_id_depthObject) {
 		// We only want to draw the instances that actually have sprites.
 		if (sprite_index > -1) {
-			draw_self();
+			//Only apply combat shaders to oAbilityCasters
+			if (object_is_ancestor(object_index, oAbilityCaster)){
+				script_execute(Shader_PickShader());	
+			} else {
+				draw_self();
```
Comment (nick-wilson95) [inline objects/oDepthManager/Draw_0.gml]: Eventually, every object type should be able to have shaders, I think it wouldn't hurt to run the following lines for all depthObjects.
	`draw_self();`
	`shader_reset();`
This way, we can remove the else block.

File: scripts/ds_EffectData_NEW/ds_EffectData_NEW.gml [hunk 7]
```diff
@@ -1,4 +1,4 @@
-/// @function ds_EffectData_NEW(enum_effect, num_steps, id_caster) Creates a new ds_EffectData
+/// @function ds_EffectData_NEW(enum_effect, num_steps, id_caster, shd_shader) Creates a new ds_EffectData
```
Comment (nick-wilson95) [inline scripts/ds_EffectData_NEW/ds_EffectData_NEW.gml]: prefix shd_shader with a ? to indicate that it's optional

File: scripts/ds_EffectData_NEW/ds_EffectData_NEW.gml [hunk 8]
```diff
@@ -1,4 +1,4 @@
...
 /// @param {enum} enum_effect The effect type.
 /// @param {number} num_steps The number of steps.
 /// @param {id} id_caster The instance id of the caster.
```
Comment (nick-wilson95) [inline scripts/ds_EffectData_NEW/ds_EffectData_NEW.gml]: Add param line for shd_shader

File: scripts/ds_EffectData_NEW/ds_EffectData_NEW.gml [hunk 9]
```diff
@@ -7,14 +7,20 @@
 var _enum_effect = argument[0];
 var _num_steps = argument[1];
 var _id_caster = argument[2];
-var _shd_shader = argument[3];
+
```
Comment (nick-wilson95) [inline scripts/ds_EffectData_NEW/ds_EffectData_NEW.gml]: I think it's neater to do:
 `var _shd_shader = argument_count > 3 ? argument[3] : undefined;`

File: scripts/ds_EffectData_NEW/ds_EffectData_NEW.gml [hunk 10]
```diff
@@ -7,14 +7,20 @@
...
 
 
 var _ds_EffectData_newEffectData = ds_map_create();
 
 _ds_EffectData_newEffectData[? DS_EFFECT_DATA_EFFECT] = _enum_effect;
 _ds_EffectData_newEffectData[? DS_EFFECT_DATA_STEPS] = _num_steps;
 _ds_EffectData_newEffectData[? DS_EFFECT_DATA_CASTER] = _id_caster;
-_ds_EffectData_newEffectData[? DS_EFFECT_DATA_SHADER] = _shd_shader;
+
+if (argument_count = 4) {
```
Comment (nick-wilson95) [inline scripts/ds_EffectData_NEW/ds_EffectData_NEW.gml]: Given previous comment, this if statement can be removed
(but leave the `_ds_EffectData_newEffectData[? DS_EFFECT_DATA_SHADER] = _shd_shader;`)

**Pull request merged**